mixer/pkg/adapter/test/integration.go

```diff
@@ -0,0 +1,270 @@
+	TestCase struct {
+		// Cfgs is a list of CRDs that Mixer will read.
+		Cfgs []string
+		// ParallelCalls is a list of test calls to be made to Mixer
+		// in parallel.
+		ParallelCalls []Call
+		// Expected json string for a WantObject object.
+		//
+		// New test can start of with an empty "{}" string and then
+		// get the baseline from the failure logs upon execution.
+		Want string
+	}
+	// Call represents the input to make a call to Mixer
+	Call struct {
+		// CallKind can be either CHECK or REPORT
+		CallKind CallKind
+		// Attrs to call the Mixer with.
+		Attrs map[string]interface{}
+		// Quotas info to call the Mixer with.
+		Quotas map[string]istio_mixer_v1.CheckRequest_QuotaParams
+	}
+	// CallKind represents the call to make; check or report.
+	CallKind int32
+
+	// Result represents the test baseline
+	Result struct {
+		// AdapterState represents adapter specific baseline data
+		AdapterState interface{} `json:"AdapterState"`
+		// Returns represents the return data from calls to Mixer
+		Returns []Return `json:"Returns"`
+	}
+	// Return represents the return data from a call to Mixer
+	Return struct {
+		// Check is the response from a check call to Mixer
+		Check adapter.CheckResult `json:"Check"`
+		// Quota is the response from a check call to Mixer
+		Quota map[string]adapter.QuotaResult `json:"Quota"`
+		// Error is the error from call to Mixer
+		Error error `json:"Error"`
+	}
+)
+
+const (
+	// CHECK for  Mixer Check
+	CHECK CallKind = 0
+	// REPORT for  Mixer Report
+	REPORT CallKind = 1
+)
+
+type (
+	setupFn    func() (ctx interface{}, err error)
+	teardownFn func(ctx interface{})
+	getStateFn func(ctx interface{}) (interface{}, error)
+)
+
+// AdapterIntegrationTest executes the given test case on
+// an in-memory Mixer with given adapters, templates and configs.
+func AdapterIntegrationTest(
+	t *testing.T,
+	adapters []adapter.InfoFn,
+	tmpls map[string]template.Info,
+	setup setupFn,
+	teardown teardownFn,
+	getState getStateFn,
+	testCase TestCase,
+
+) {
+
+	// Let the test do some initial setup.
+	var ctx interface{}
+	var err error
+	if setup != nil {
+		ctx, err = setup()
+		// Teardown the initial setup
+		if teardown != nil {
+			defer teardown(ctx)
+		}
+		if err != nil {
+			t.Fatalf("initial setup failed: %v", err)
+		}
+	}
+
+	// Start Mixer
+	var args *server.Args
+	var env *server.Server
+	if args, err = getServerArgs(tmpls, adapters, testCase.Cfgs); err != nil {
+		t.Fatalf("fail to create mixer args: %v", err)
+	} else if env, err = server.New(args); err != nil {
+		t.Fatalf("fail to new mixer: %v", err)
+	} else {
+		env.Run()
+		defer closeHelper(env)
+	}
+
+	// Connect the client to Mixer
+	conn, err := grpc.Dial(env.Addr().String(), grpc.WithInsecure())
+	if err != nil {
+		t.Fatalf("Unable to connect to gRPC server: %v", err)
+	}
+	client := istio_mixer_v1.NewMixerClient(conn)
+	defer closeHelper(conn)
+
+	// Invoke calls async
+	done := make(chan Return)
```

Comment:
should we use a `sync.WaitGroup` here?

Reply:
Absolutely, few less lines of code. Good catch